cmd/zoekt-sourcegraph-indexserver/cleanup.go

```diff
@@ -283,3 +283,79 @@ func shardsLog(indexDir, action string, shards []shard, repoName string) {
 		_, _ = fmt.Fprintf(shardLogger, "%d\t%s\t%s\t%d\t%s\n", time.Now().UTC().Unix(), action, shard, shardSize, repoName)
 	}
 }
+
+// vacuum removes tombstoned repos from compound shards. Ensure vacuum has
+// exclusive access to dir while it runs.
+func vacuum(dir string) {
+	d, err := os.Open(dir)
+	if err != nil {
+		return
+	}
+	defer d.Close()
+	fns, _ := d.Readdirnames(-1)
+
+	for _, fn := range fns {
+		// We could run this over all shards, but based on our current setup, simple
+		// shards won't have tombstones but instead will be moved to .trash.
+		if !strings.HasPrefix(fn, "compound-") || !strings.HasSuffix(fn, ".zoekt") {
+			continue
+		}
+
+		removed, err := removeTombstones(filepath.Join(dir, fn))
+		if err != nil {
+			debug.Printf("error while removing tombstones in %s: %s", fn, err)
+		}
+		if len(removed) > 0 {
+			for _, ts := range removed {
+				shardsLog(dir, "vac", []shard{{
+					Path: filepath.Join(dir, fn)}}, ts)
+			}
+		}
+	}
+}
+
+// removeTombstones removes all tombstones from a compound shard at fn by merging
+// the compound shard with itself.
+func removeTombstones(fn string) ([]string, error) {
+	repos, _, err := zoekt.ReadMetadataPath(fn)
+	if err != nil {
+		return nil, fmt.Errorf("zoekt.ReadMetadataPath: %s", err)
+	}
+
+	var tombstones []string
+	for _, r := range repos {
+		if r.Tombstone {
+			tombstones = append(tombstones, r.Name)
+		}
+	}
+	if len(tombstones) == 0 {
+		return nil, nil
+	}
+
+	f, err := os.Open(fn)
+	if err != nil {
+		return nil, fmt.Errorf("os.Open: %s", err)
+	}
+	defer f.Close()
+
+	indexFile, err := zoekt.NewIndexFile(f)
+	if err != nil {
+		return nil, fmt.Errorf("zoekt.NewIndexFile: %s ", err)
+	}
+	defer indexFile.Close()
+
+	defer func() {
+		paths, err := zoekt.IndexFilePaths(fn)
+		if err != nil {
+			return
+		}
+		for _, path := range paths {
+			os.Remove(path)
+		}
+	}()
+	_, err = zoekt.Merge(filepath.Dir(fn), indexFile)
+	if err != nil {
+		return nil, fmt.Errorf("zoekt.Merge: %s", err)
+	}
+	return tombstones, nil
+}
```

Comment:
rather than merging in here, can we do it in a subprocess like we do for other jobs? I worry about how this will mess with GC/etc. Also quite nice to think of this process as only ever supervising, never doing heavy lifting.

Reply:
Done. Now we call out to zoekt-merge-index instead. I had to update the tests to mock the system call which bloats the diff a bit.